### ISSUE TITLE
Rigidifying a large protein

### ISSUE THREAD

**Issue by tanmoy7989** [openmm/openmm#3454]
Is there a simple way to do the following?

I have a system with a rigid protein - flexible ligand (small molecule) in implicit water, and I want to study the possible "epitopes" of the small molecule on the protein (not too interested in the thermodynamics at this stage). So, I want to "rigidify" the protein and to that end I added a `CustomExternalForce` (essentially a very stiff spring) that restrains each protein atom to its original position. However, I am thinking that for further speed-up I would remove the intra-protein energy terms. Admittedly this could mess up the thermodynamics, but like I said, I don't care about thermodynamics in this prelim investigation.  However, I do want the protein-ligand interactions all intact. 

From what I understand (I'm completely new to openMM), I could track the `NonBondedForce`  terms between protein atoms and add each of them as an `Exception` with `chargeProd=0` and `epsilon=0`.  But I can't do this with bonded, angle or torsion terms. Note that I don't want to *turn off* these forces by zeroing their parameters, I want to completely remove them from calculation. 

Is there a more openMM-recommended way to do this?

**Comment by peastman** [comment]
> So, I want to "rigidify" the protein and to that end I added a CustomExternalForce (essentially a very stiff spring) that restrains each protein atom to its original position.

An even easier way is to set the masses of the protein atoms to 0.  If an atom is massless, the integrator will never attempt to move it.

> Note that I don't want to turn off these forces by zeroing their parameters, I want to completely remove them from calculation.

There's no really clean way to do that.  We probably should add `removeBond()`, `removeAngle()`, etc. methods to the relevant forces.  But for the moment, here's some code I wrote in the OpenMM-ML package that does exactly what you're describing:

https://github.com/openmm/openmm-ml/blob/4f0546678eeb9b2703d28d3fce7f678802640701/openmmml/mlpotential.py#L223-L274

It does it in a slightly obscure way: it serializes the System to XML, then edits the XML representation to remove any bond, angle, or torsion for the unwanted atoms, then deserializes it to construct a new System.

**Comment by tanmoy7989** [comment]
Thanks @peastman! If I zero out the masses of the protein atoms (and don't edit the XML), intra-protein forces would still be calculated and that would be wasteful for my purposes, correct?

**Comment by peastman** [comment]
Correct.  Although it's probably not that big a deal.  Nonbonded forces usually take a lot more time than bonded ones.  But there would still be some difference in speed.

**Comment by peastman** [comment]
That's probably because of the nonbonded exceptions.  They're a lot more expensive to compute than regular interactions.  If the atoms are massless, you can skip adding them.  The internal forces between protein atoms won't affect anything.

**Comment by tanmoy7989** [comment]
I see. So this doesn't apply then?
![image](https://user-images.githubusercontent.com/8772756/153271628-de2b7921-2abe-466e-8554-598664d98276.png)

**Comment by tanmoy7989** [comment]
Thanks again @peastman. This makes sense, and I was able to bring down the runtime by ~ 30% by turning off only the bonds, angles, torsions and constraints (and zeroing the masses).

**Comment by tanmoy7989** [comment]
@peastman One last (hopefully!) question: 

I was able to rigidify the protein by turning off the intra-protein bond, angle and torsions and further setting all protein atom masses to zero, so that protein atoms don't participate in time integration any more. However, I needed to tether one part of the ligand to a given protein atom, because I know that they stick together. The only way I was able to do this was to "un-rigidify" the participating protein atom (nonzero mass and bond / angle / torsion force on this guy). That got me thinking: When I rigidified the protein, specifically when I turned off the masses, do the protein atoms still contribute to nonbonded forces on the ligand atoms? Because otherwise, the whole binding simulation would be pointless. Sorry, this is probably a stupid question, but better to get it clarified before running longer simulations.
